File: newbyschwemmer/Domain.cpp
//
// Created by konstantin on 4/7/20.
//
#include <cmath>

#include "Domain.h"

/*
 * Domain
 */
Domain::Domain(double rho_min, double rho_max) {
    if (rho_max >= rho_min) {
        rho_UB = rho_max;
        rho_LB = rho_min;
    } else
        {
        rho_LB = rho_max;
        rho_UB = rho_min;
    }
    width = rho_UB - rho_LB;
}

bool Domain::hit_LowerBoundary(std::array<double, 2>& pos) {
    return (pos[0] < rho_LB);
}

bool Domain::hit_UpperBoundary(std::array<double, 2>& pos) {
    return (pos[0] > rho_UB);
}

std::array<double, 2> Domain::reflect_at_LowerBoundary(std::array<double, 2>& pos) {
    // pos = [rho, phi]
    double Drho = abs(pos[0] - rho_LB);
    int m = floor(Drho / width);
    if (m % 2 == 0){
        return {rho_LB + (Drho - m*width), pos[1]};
    } else{
        return {rho_UB - (Drho - m*width), pos[1]};
    }
}

std::array<double, 2> Domain::reflect_at_UpperBoundary(std::array<double, 2>& pos) {
    // pos = [rho, phi]
    double Drho = abs(pos[0] - rho_UB);
    int m = floor(Drho / width);
    if (m % 2 == 0){
        return {rho_UB - (Drho - m*width), pos[1]};
    } else{
        return {rho_LB + (Drho - m*width), pos[1]};
    }
}

std::array<double, 2> Domain::apply(std::array<double, 2>& pos) {
    if (this->hit_LowerBoundary(pos)){
        return this->reflect_at_LowerBoundary(pos);
    } else{
        return this->reflect_at_UpperBoundary(pos);
    }
}

File: PyTools/Toolbox/Domain.py
import numpy as np


class Domain():

    def apply_boundary_conditions(self, x_init: np.array, x_final: np.array) -> np.array:
        pass


class ReflectiveAnnulus(Domain):

    def __init__(self, rho_min: float, rho_max: float):
        if (rho_max < rho_min):
            self.rho_min = rho_max
            self.rho_max = rho_min
        else:
            self.rho_min = rho_min
            self.rho_max = rho_max

        self.width = np.abs(rho_max - rho_min)


    def apply_boundary_conditions(self, x_init: np.array, x_final: np.array) -> np.array:
        if (x_final[0] < self.rho_min):
            Drho = np.abs(x_final[0] - self.rho_min)
            m = np.floor(Drho / self.width)
            if ((m % 2) == 0):
                return np.array([self.rho_min + (Drho - m*self.width), x_final[1]])
            else:
                return np.array([self.rho_max - (Drho - m*self.width), x_final[1]])
        else:
            if (x_final[0] > self.rho_max):
                Drho = np.abs(x_final[0] - self.rho_max)
                m = np.floor(Drho / self.width)
                if ((m % 2) == 0):
                    return np.array([self.rho_max - (Drho - m*self.width), x_final[1]])
                else:
                    return np.array([self.rho_min + (Drho - m*self.width), x_final[1]])
            else:
                return x_final